### ISSUE TITLE
Additional address(es) in access.log

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7404]
|<img alt="jwashin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jwashin reported|
|-|-|
|Trac ID|trac#1468|
|Type|enhancement|
|Created|2006-02-04 00:26:11Z|
|Branch|https://github.com/twisted/twisted/tree/x-forwarded-for-logging-1468|

Attachments:

* [x-forwarded-for_in_log.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/742/742f0a1ef06d6bdc8c856b05b3ad8839e2a27ba7/45a04c5e4b671bae3e832930fc4e62f2749a1139.diff) (1401 bytes) - added by jwashin on 2006-02-04 00:26:10Z - 
* [x-forwarded-for_in_web_http.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/742/742f0a1ef06d6bdc8c856b05b3ad8839e2a27ba7/9bbde7842514ae61a9ffde0bc1968ae4ee47a281.patch) (3278 bytes) - added by damnwidget on 2013-04-20 03:33:31Z - X-Forwarded-For in twisted.web.http.HTTPFactory.log
* [x-forwarded-for-logging-1468-example.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/742/742f0a1ef06d6bdc8c856b05b3ad8839e2a27ba7/8741768b3906708f54ae42476def16aeda69feab.patch) (2096 bytes) - added by rwall on 2014-01-29 17:45:58Z - The example script I used to test the functionality in this branch. Along with some necessary changes to t.w.proxy module and example so that I could use twisted as the frontend proxy.

<details><summary>Searchable metadata</summary>

```
trac-id__1468 1468
type__enhancement enhancement
reporter__jwashin jwashin
priority__normal normal
milestone__ 
branch__branches_x_forwarded_for_logging_1468 branches/x-forwarded-for-logging-1468
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1139012771000000 1139012771000000
changetime__1391653743000000 1391653743000000
version__ 
owner__exarkun exarkun
cc__jknight cc__jwashin cc__exarkun cc__akrherz
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jwashin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jwashin commented|
|-|-|

```
#!html
<pre>
Attached diff against twisted.web2.log.py for adding additional host (e.g.,
proxy) addresses (besides 127.0.0.1 or 192.168.x.x) to the access log if twisted
runs behind a reverse proxy.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to jwatt|
|-|-|

jwashin, can you do one or two things with this?

  * Add unit tests
  * Port it to `twisted.web`, where web-related development effort is currently focused?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I'm going to implement this by having a site wrapper.  It will have a `getResourceFor` method which returns a resource wrapper which intercepts the request passed to `render` and wraps it in a request wrapper which looks at `X-Forwarded-For` and uses its value to decide what to return from `getClientIP`.

Another thing which would be nice is if Twisted's proxy would set the `X-Forwarded-For` header.  These two changes together would be a good improvement to logging and such for Twisted web servers (in particular, I am excited to have our own trac and buildbot servers logging correct client address information as a result of this).

**Comment by twisted-trac** [comment]
|<img alt="akrherz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|akrherz commented|
|-|-|

I stumbled into this ticket today, right or probably wrong, I ended up adding this simple log statement in my render() 
```
  log.msg('%s %s' % (request.uri, request.getHeader('x-forwarded-for')))
```

I am using the old twisted.web ,thanks

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="damnwidget's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|damnwidget commented|
|-|-|

Hi

I need this behaviour in my project so I wrote a patch to make it available. My way is not as sophisticated as Jean-Paul suggested but it works.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Unfortunately, this patch breaks `getClientIP` for any existing users which thought it would return an IP address.

**Comment by twisted-trac** [comment]
|<img alt="damnwidget's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|damnwidget commented|
|-|-|

You are right.

We can always just split the string returned by getHeader and return the last inserted value to the string as the method should return always a single IP address.

Or maybe we should just focus in JP approach to solve it and discard this simplistic way of try to fix this.

What do you think?

Regards.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

A more limited solution (one that doesn't fix `getClientIP` in general, just the log file) would be to make `Site` configurable about whether it logs `getClientIP` or the value of the `X-Forwarded-For` header.  This would probably be a lot easier and it's better than nothing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [41345]) Branching to 'x-forwarded-for-logging-1468'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

I implemented that more limited solution (which was quite sufficiently hard all on its own).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @exarkun<br>@wallrj set status to `new`|
|-|-|

Thanks exarkun.

This looks useful.

Notes:
 1. Merges cleanly

Points:
 1. General
    1. It wasn't clear how the new log formatters are meant to be used.
    2. Maybe add some brief documentation and an example.
    3. Can they be supplied to twistd somehow?
       1. No, it seems not. maybe raise a ticket for supplying a custom  access log formatter to the twistd web plugin.
    4. twistd web --help talks about the "Combined Log Format" https://httpd.apache.org/docs/1.3/logs.html#combined
       1. And that's what your "commonLogFormat" function seems to produce.
       2. "Common Log Format" shouldn't have the referrer and user agent according to https://httpd.apache.org/docs/1.3/logs.html#common
    5. I've attached my simple example as a patch (along with some other changes which were necessary to make it work)

 2. Build Results:
    1. Twistedchecker
       1. Some long lines in test_web, but I see those are example log lines, and it's easier to read them in one line.
    2. Pydoctor
       1. "twisted.web.iweb.IAccessLogFormatter.__call__: invalid ref to Request"
       2. https://buildbot.twistedmatrix.com/builders/documentation/builds/4368/steps/api-documentation/logs/new%20pydoctor%20errors

 3. [branches/x-forwarded-for-logging-1468/twisted/web/iweb.py](branches/x-forwarded-for-logging-1468/twisted/web/iweb.py)
    1. IAccessLogFormatter
       1. "timestamp: L{unicode}" is it really unicode, or a datetime instance?

 4. [branches/x-forwarded-for-logging-1468/twisted/web/test/test_web.py](branches/x-forwarded-for-logging-1468/twisted/web/test/test_web.py)
    1. "reactor.advance(1234567890)" -- why 1234567890?
    2. test_xforwardedfor
       1. [b"172.16.1.2, 10.0.0.3, 192.168.1.4"] ...is that right? Or should those IPs be separate list items?
    3. assertLogs
       1.
    4. testSimple, testMethodQuote, testRequestQuote
       1. Consider standardizing the names of these old test methods and adding docstrings where necessary.

 5. [branches/x-forwarded-for-logging-1468/twisted/web/http.py](branches/x-forwarded-for-logging-1468/twisted/web/http.py)
    1. Typo "See the C{logFormatter} parameter to L{__init__}"
    2. commonLogFormatter
       1. Add an @see reference to the interface so that people can see the parameter documentation.
       2. Maybe file a Pydoctor ticket that it should recognise and link to
          interface documentation for functions that directlyProvide and
          Interface
       3. Maybe file a Pydoctor ticket that the Interface "Known
          Implementations" links should be created for functions which directly
          provide an interface.
    3. _XForwardedForRequest
       1. You use the first element in the comma separated x-forwarded-for
          header, but what if there is a chain of proxies. Do you definately
          want the first? Or do you want the last?
          1. I can't tell from the wikipedia doc
          2. Maybe double check and add a link to definitive documentation of x-forwarded-for
             1. Maybe this http://tools.ietf.org/html/draft-ietf-appsawg-http-forwarded-10
             2. Although this suggests that there may not be a standard: https://stackoverflow.com/questions/13111080/what-is-a-full-specification-of-x-forwarded-proto-http-header
          3. Check how Apache or Nginx do it.
             1. Nginx says it uses the last host
             2. http://nginx.org/en/docs/http/ngx_http_realip_module.html
       2. Also, wikipedia says that the hosts may be separated with comma and
          space, but you only split on comma. Perhaps you should strip the
          host before logging it.
       3. Also I wonder if the host may ever be "host:port" so it may not be sensible
          to call the method "getClientIP" unless you remove the port if
          present.
          1. https://en.wikipedia.org/wiki/X-Forwarded-For
          2. I did some checking and it seems that it is only ever the host IP.

 6. [branches/x-forwarded-for-logging-1468/twisted/web/topfiles/1468.feature](branches/x-forwarded-for-logging-1468/twisted/web/topfiles/1468.feature)
    1. You say that the new logFormatter function is available in subclasses,
       but actually you haven't added it to server.Site.
    2. So there's no way to create a site with a proxiedLogFormatter.
    3. Maybe this branch should make that change to the server.Site constructor (with a test or two)


Please answer or address the numbered points above and resubmit for review.

And if you feel like reviewing [#6862](https://github.com/twisted/twisted/issues/6862) I'd be most grateful :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner|
|-|-|

Thanks for the thorough review!

> IAccessLogFormatter - "timestamp: L{unicode}" is it really unicode, or a datetime instance?

It's really `unicode` :(  This is not a great API.  I could try to fix it... or at least talk about it.  I left this alone because Twisted Web goes out of its way to not format the timestamp into a string more than once per second (because long ago someone noticed this string formatting was the bottleneck for Twisted Web).

> Typo `See the C{logFormatter} parameter to L{__init__}`

I don't see it.

> Maybe file a Pydoctor ticket that it should recognise and link to interface documentation for functions that directlyProvide and Interface

I switched this to `provider` which should be easier for pydoctor to recognize.

https://bugs.launchpad.net/pydoctor/+bug/1274353

> You use the first element in the comma separated x-forwarded-for header, but what if there is a chain of proxies. Do you definately want the first? Or do you want the last?

I thought I wanted the first.  The docs I read said the first is the "most trustworthy" (or some such) because proxies add addresses to the *beginning*.

I guess maybe I'm not massively concerned by this because the most common scenario is a single level of proxying and the other half of this branch makes it possible to supply a formatter with different logic, if you want.

Reading the nginx documentation, I think what that extension does is actually like what's implemented in the branch.  It does talk about "last" a lot - but it's talking about interpreting the list in conjunction with another list of trusted proxy addresses.  So in effect it is selecting the *right-most* *trusted* address.  Since this branch doesn't have the notion of trusted addresses, the left-most address is the right-most trusted address.  I could imagine extending this feature to be as flexible as the nginx extension someday.

I think I addressed all the rest of the review points.  Sorry I didn't call out the revisions fixing each point.  I found it a little hard to do that because of the combination of the nesting in the review formatting and the multiple related bullet points for a number of things.  That wasn't a problem otherwise, just thought I'd mention it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @exarkun|
|-|-|

Sorry, there are Python 3 failures now, don't have time to look at them now, maybe soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun removed owner|
|-|-|

Fixed those by skipping the failing tests on Python 3.  Yay.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Thanks exarkun,

Notes:
 1. Tested the code using a hacked twisted reverse proxy -- it works -- even with multiple chained proxies.
 2. I only reviewed the changes since my previous review.

Points:
 1. Build failures
    1. Windows line ending failures: https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/3586/steps/select/logs/problems
 2. [branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/access-logging.rst](branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/access-logging.rst)
    1. Typo "Twisted Web come with one"
    2. Typo "for an example that both these features"
    3. Syntax error in the final example "proxiedLogFormatter))"
 3. [branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/index.rst](branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/index.rst)
    1. I think the toctree is important -- it seems to be required by sphinx for its breadcrumb navigation. See [#6940](https://github.com/twisted/twisted/issues/6940) and ticket:6905#comment:8
    2. It's a guess, but the way you've formatted the list of articles means it will be neither bulletted nor numbered. Is that intentional? I'd rather see bullet points personally.

Please merge after fixing the numbered points above.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

Thanks for the review!

> [branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/access-logging.rst](branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/access-logging.rst)

Fixed in r41551

> Windows line ending failures

Fixed in r41552.

> [branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/index.rst](branches/x-forwarded-for-logging-1468/docs/projects/web/howto/web-in-60/index.rst)

I'm stumped.  I don't understand the rules or requirements for this file.  Maybe I should revert all of the changes and *only* add the new document to both places?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

[changeset:41562 I] [changeset:41563 punted].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `closed`|
|-|-|

(In [41565]) Merge x-forwarded-for-logging-1468 

Author: exarkun
Reviewer: rwall
Fixes: [#7404](https://github.com/twisted/twisted/issues/7404)

Add support to Twisted Web for customizing the lines written to the server's
access log.  Also add a helper which uses this new feature to write access
lines that respect the X-Forwarded-For header to improve access logs when
deployed behind a proxy of some sort.
